File: backend/app/schemas/engagement.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime, date

# -----------------------------------------
# Executive Engagement Overview & Activity
# -----------------------------------------

class ActiveUsersOverview(BaseModel):
    dau: int = Field(..., description="Daily Active Users")
    wau: int = Field(..., description="Weekly Active Users")
    mau: int = Field(..., description="Monthly Active Users")
    stickiness_ratio: float = Field(..., description="DAU / MAU ratio (%)")

class ActivityHeatmapPoint(BaseModel):
    hour: int = Field(..., ge=0, le=23)
    day_of_week: int = Field(..., ge=0, le=6)
    active_users: int

class ActivityResponse(BaseModel):
    time_slot_breakdown: Dict[str, int] = Field(
        ..., description="Counts for Morning, Afternoon, Evening, Night"
    )
    heatmap_points: List[ActivityHeatmapPoint] = Field(default_factory=list)

# -----------------------------------------
# Session Intelligence
# -----------------------------------------

class SessionMetricsResponse(BaseModel):
    total_sessions: int
    avg_duration_minutes: float
    longest_session_minutes: float
    sessions_per_user: float
    device_breakdown: Dict[str, int] = Field(default_factory=dict)
    platform_breakdown: Dict[str, int] = Field(default_factory=dict)

# -----------------------------------------
# Listening Intelligence
# -----------------------------------------

class ListeningIntelligenceResponse(BaseModel):
    total_listening_hours: float
    avg_listening_hours_per_user: float
    completion_rate: float = Field(..., description="Track completion percentage")
    skip_rate: float = Field(..., description="Track skip percentage (<30s)")
    replay_rate: float = Field(..., description="Track replay percentage")
    songs_per_session: float
    albums_per_session: float
    artists_per_session: float

# -----------------------------------------
# Feature Adoption
# -----------------------------------------

class FeatureAdoptionMetric(BaseModel):
    feature_name: str
    category: str = Field(..., description="Playback, Discovery, Social, Premium")
    adoption_rate: float
    growth_rate: float
    retention_rate: float

class FeatureAdoptionResponse(BaseModel):
    features: List[FeatureAdoptionMetric] = Field(default_factory=list)

# -----------------------------------------
# User Journey & Funnels
# -----------------------------------------

class FunnelStep(BaseModel):
    step_number: int
    step_name: str
    user_count: int
    conversion_rate: float
    dropoff_rate: float

class UserJourneyResponse(BaseModel):
    funnel_name: str
    total_started: int
    total_completed: int
    overall_conversion_rate: float
    steps: List[FunnelStep] = Field(default_factory=list)

# -----------------------------------------
# Behavior Segmentation
# -----------------------------------------

class BehaviorSegmentPoint(BaseModel):
    segment_name: str = Field(..., description="e.g., Casual, Power, Explorer, Binge")
    user_count: int
    percentage: float
    description: Optional[str] = None

class BehaviorSegmentationResponse(BaseModel):
    total_users_classified: int
    segments: List[BehaviorSegmentPoint] = Field(default_factory=list)

# -----------------------------------------
# Engagement Health Engine
# -----------------------------------------

class SubScoreBreakdown(BaseModel):
    activity_score: float
    volume_score: float
    session_quality_score: float
    feature_breadth_score: float

class EngagementHealthResponse(BaseModel):
    overall_engagement_score: float = Field(..., description="0-100 normalized score")
    health_category: str = Field(..., description="Excellent, Good, Average, Poor, Critical")
    sub_scores: SubScoreBreakdown

File: frontend/src/api/engagement.ts
import axios from 'axios';
import { StandardAnalyticsResponse, ExecutiveInsight } from './retention';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1';

export interface ActiveUsersOverview {
  dau: number;
  wau: number;
  mau: number;
  stickiness_ratio: number;
}

export interface ActivityHeatmapPoint {
  hour: number;
  day_of_week: number;
  active_users: number;
}

export interface ActivityResponse {
  time_slot_breakdown: Record<string, number>;
  heatmap_points: ActivityHeatmapPoint[];
}

export interface SessionMetricsResponse {
  total_sessions: number;
  avg_duration_minutes: number;
  longest_session_minutes: number;
  sessions_per_user: number;
  device_breakdown: Record<string, number>;
  platform_breakdown: Record<string, number>;
}

export interface ListeningIntelligenceResponse {
  total_listening_hours: number;
  avg_listening_hours_per_user: number;
  completion_rate: number;
  skip_rate: number;
  replay_rate: number;
  songs_per_session: number;
  albums_per_session: number;
  artists_per_session: number;
}

export interface FeatureAdoptionMetric {
  feature_name: string;
  category: string;
  adoption_rate: number;
  growth_rate: number;
  retention_rate: number;
}

export interface FeatureAdoptionResponse {
  features: FeatureAdoptionMetric[];
}

export interface FunnelStep {
  step_number: number;
  step_name: string;
  user_count: number;
  conversion_rate: number;
  dropoff_rate: number;
}

export interface UserJourneyResponse {
  funnel_name: string;
  total_started: number;
  total_completed: number;
  overall_conversion_rate: number;
  steps: FunnelStep[];
}

export interface BehaviorSegmentPoint {
  segment_name: string;
  user_count: number;
  percentage: number;
  description?: string;
}

export interface BehaviorSegmentationResponse {
  total_users_classified: number;
  segments: BehaviorSegmentPoint[];
}

export interface SubScoreBreakdown {
  activity_score: number;
  volume_score: number;
  session_quality_score: number;
  feature_breadth_score: number;
}

export interface EngagementHealthResponse {
  overall_engagement_score: number;
  health_category: string;
  sub_scores: SubScoreBreakdown;
}

// -----------------------------------------
// API Fetchers
// -----------------------------------------

export const fetchEngagementOverview = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<ActiveUsersOverview>> => {
  const res = await axios.get(`${API_BASE_URL}/engagement/overview`, { params });
  return res.data;
};

export const fetchActivityBreakdown = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<ActivityResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/engagement/activity`, { params });
  return res.data;
};

export const fetchSessionMetrics = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<SessionMetricsResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/engagement/sessions`, { params });
  return res.data;
};

export const fetchListeningMetrics = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<ListeningIntelligenceResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/engagement/listening`, { params });
  return res.data;
};

export const fetchFeatureAdoption = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<FeatureAdoptionResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/engagement/features`, { params });
  return res.data;
};

export const fetchFunnelAnalysis = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<UserJourneyResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/engagement/funnels`, { params });
  return res.data;
};

export const fetchBehaviorSegments = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<BehaviorSegmentationResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/engagement/segments`, { params });
  return res.data;
};

export const fetchEngagementHealth = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<EngagementHealthResponse>> => {
  const res = await axios.get(`${API_BASE_URL}/engagement/health`, { params });
  return res.data;
};

export const fetchEngagementInsights = async (params?: Record<string, any>): Promise<StandardAnalyticsResponse<ExecutiveInsight[]>> => {
  const res = await axios.get(`${API_BASE_URL}/engagement/insights`, { params });
  return res.data;
};
